native/src/main/java/io/ballerina/stdlib/serdes/Deserializer.java

```diff
@@ -52,251 +66,335 @@
-        } else {
-            throw createSerdesError(UNSUPPORTED_DATA_TYPE + type.getName(), SERDES_ERROR);
+            case TypeTags.UNION_TAG: {
+                return getUnionTypeValueFromMessage(dynamicMessage, ballerinaType);
+            }
+
+            case TypeTags.ARRAY_TAG: {
+                Type elementType = ((ArrayType) ballerinaType).getElementType();
+                Descriptor messageDescriptor = dynamicMessage.getDescriptorForType();
+                FieldDescriptor fieldDescriptor = messageDescriptor.findFieldByName(ARRAY_FIELD_NAME);
+                Object messageFieldValue = dynamicMessage.getField(fieldDescriptor);
+                return getArrayTypeValueFromMessage(messageFieldValue, elementType, messageDescriptor);
+            }
+
+            case TypeTags.RECORD_TYPE_TAG: {
+                return getRecordTypeValueFromMessage(dynamicMessage, (RecordType) ballerinaType);
+            }
+
+            default:
+                throw createSerdesError(UNSUPPORTED_DATA_TYPE + ballerinaType.getName(), SERDES_ERROR);
         }
     }
 
-    private static Object getBallerinaPrimitiveValueFromMessage(Object value) {
+    private static Object getDecimalPrimitiveTypeValueFromMessage(DynamicMessage decimalMessage) {
+
+        Descriptor decimalSchema = decimalMessage.getDescriptorForType();
+
+        FieldDescriptor valueField = decimalSchema.findFieldByName(VALUE);
+        FieldDescriptor scaleField = decimalSchema.findFieldByName(SCALE);
+        FieldDescriptor precisionField = decimalSchema.findFieldByName(PRECISION);
+
+        BigInteger value = new BigInteger(((ByteString) decimalMessage.getField(valueField)).toByteArray());
+        Integer scale = (Integer) decimalMessage.getField(scaleField);
+        MathContext precision = new MathContext((int) decimalMessage.getField(precisionField));
+
+        BigDecimal bigDecimal = new BigDecimal(value, scale, precision);
+        return ValueCreator.createDecimalValue(bigDecimal);
+    }
+
+    private static Object getPrimitiveTypeValueFromMessage(Object value) {
+
+        if (value instanceof DynamicMessage) {
+            DynamicMessage decimalMessage = ((DynamicMessage) value);
+            return getDecimalPrimitiveTypeValueFromMessage(decimalMessage);
+        }
+
+        if (value instanceof ByteString) {
+            return ((ByteString) value).byteAt(0);
+        }
+
         if (value instanceof String) {
             return StringUtils.fromString((String) value);
         }
+
         return value;
     }
 
-    private static Object getBallerinaArrayValueFromMessage(Object value, Type type, Descriptor schema,
-                                                            int unionFieldIdentifier) {
+
+    private static Object getUnionTypeValueFromMessage(DynamicMessage dynamicMessage, Type type) {
+
+        Descriptor messageDescriptor = dynamicMessage.getDescriptorForType();
+
+        for (var entry : dynamicMessage.getAllFields().entrySet()) {
```

Comment:
Let's not use `var` until it is absolutely needed.

Reply:
Updated accordingly